### ISSUE TITLE
HelmRelease targetNamespace not honored by helm release secret

### ISSUE THREAD

**Issue by tommasopozzetti** [fluxcd/helm-controller#185]
I have noticed a strange behavior when using the targetNamespace feature of the HelmRelease.
My understanding was that installing a HelmRelease resource in `targetNs` or installing it in `hrNs` with targetNamespace `targetNs` should result in the same behavior from helm's point of view, with the only change being in where the helm-controller stores metadata to manage the reconciliation.
However I am seeing that when installing a HelmRelease in `hrNs` with targetNamespace `targetNs`, helm is somehow set up to store the release's secrets in `hrNs` rather than `targetNs` even though the release indeed targets `targetNs`.
This results in the weird behavior of that helm release not showing up when running `helm list` in `targetNs` but showing up with a different namespace when running `helm list` in `hrNs` as follows (output is simplified for better understanding):
```
$ helm list -n targetNs
NAME    NAMESPACE       REVISION        UPDATED STATUS  CHART   APP VERSION

$ helm list -n hrNs
NAME                            NAMESPACE       REVISION        UPDATED                                 STATUS        CHART                    APP VERSION
helmRelease                     hrNs            6               2020-12-28 16:47:12.134388 -0600 CST    deployed      helmRelease-0.0.12
helmRelease2                    hrNs            4               2020-12-22 15:31:46.448803 -0600 CST    deployed      helmRelease2-0.1.0 
fluxRelease                     targetNs        14              2020-12-28 21:56:58.786390024 +0000 UTC deployed      fluxRelease-0.7.4
```
In the previous example, the two helmRelease helm releases are manually installed via helm while fluxRelease has been installed by the helm-controller.

This behavior also brings as consequence the fact that if `fluxRelease` had already been manually installed via helm and/or previously managed by helm-operator, when helm-controller starts the reconciliation process for the first time it believes that it is not yet installed (since the secret would live in `targetNs` and therefore it tries to do an install which might leave everything in a weird state considering the release already exists.

This behavior, as far as I have found, is also hard to reproduce using helm directly, as I don't see any option to install a release storing its secret in a different namespace, which might beg the question whether it will actually even remain supported by helm.

I think the reason for this is the initialization of the helm client [here](https://github.com/fluxcd/helm-controller/blob/v0.4.4/controllers/helmrelease_controller.go#L263) using the namespace of the HelmRelease rather than the targetNamespace.

Is this behavior intended? If so, is it officially supported by helm and do you have any advice on how to migrate a helm release already installed without having to purge it and re-install it?

**Comment by seaneagan** [comment]
It is intentional, there is some discussion in #96. I also recall some discussion about wanting to implement a flag for this upstream in the Helm CLI, but can't find that.

**Comment by tommasopozzetti** [comment]
@seaneagan Thank you for the info!
The issue I see with this approach is that, besides not being able to manually upgrade the helm release as mentioned in the discussion you linked, the `targetNamespace` used to work differently with helm-operator. Therefore, if users are migrating from helm-operator using `targetNamespace` to helm-controller, the migration guide is no longer valid.
Is there any plan to possibly make the storage namespace for `targetNamespace` configurable?
And is there any official guide/advice for this specific migration?

**Comment by CuBiC3D** [comment]
I just stumbled upon that in combination with `kubeConfig` by trying to deploy a `HelmRelease` on another cluster. It fails because the namespace where the `HelmRelease` is applied does not exist on the remote cluster. This breaks the remote cluster ability.
```
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: my-chart
  namespace: local-namespace
spec:
  interval: 5m
  releaseName: my-chart
  targetNamespace: default
  kubeConfig:
    secretRef:
      name: kubeconfig
  chart:
    spec:
      chart: my-chart
      version: 1.0.0
      sourceRef:
        kind: HelmRepository
        name: my-repo
        namespace: flux-system
```
```
{"level":"error","ts":"2021-01-07T22:49:05.248Z","logger":"controller","msg":"Reconciler error","reconcilerGroup":"helm.toolkit.fluxcd.io","reconcilerKind":"HelmRelease","controller":"helmrelease","name":"my-chart","namespace":"local-namespace","error":"Helm install failed: create: failed to create: namespaces \"local-namespace\" not found"}
```

**Comment by gtracer** [comment]
@CuBiC3D I am running into the exact same issue with remote clusters. 

@tommasopozzetti similar to you I found that it is the [helm client config code](https://github.com/fluxcd/helm-controller/blob/220940b0444d3e24c89a135110de9fcdff58e7dd/controllers/helmrelease_controller.go#L263) that is problematic. 

For remote users it is a pretty limiting factor because that code is essentially stripping any ability to set the storage namespace. In comparison, with r[aw helm it defaults to the namespace](https://helm.sh/docs/faq/#release-names-are-now-scoped-to-the-namespace) used in my config so I can at least implicitly set it. 

I see a few possible solutions:
1. Let's treat `targetNamespace` as the namespace for both storage and manifests (the same way helm treats `--namespace`)
2. Don't override the namespace at all and instead use the namespace on the remote kubeconfig ([the same way the helm docs recommend today](https://helm.sh/docs/faq/#release-names-are-now-scoped-to-the-namespace))
3. Let's expose a second property on the HelmRelease called storageNamespace and let users pick their preferred storageNamespace

As I read through the code, I don't see any glaring reasons why we couldn't configure the storageNamespace since:
1. The runner is instantiated for every reconcile (allowing HelmReleases to have different storageNamespaces)
2. The runner seems to be nicely designed as the single layer for working with Helm therefore making the code change pretty clean 

Tagging those who seem like they have interest / knowledge based on #96 

@seaneagan @hiddeco 

Also, happy to contribute code to any solutions

**Comment by hiddeco** [comment]
The feature was initially aimed at multi-tenancy where "something" would be injected into the tenants namespace, and hiding the storage from the tenant made sense for this. Since the introduction of the `KubeConfig` feature, and because people at times seem to have other use-cases and/or structural setups, this seems to be a limitation.

To offer full flexibility, I think

> Let's expose a second property on the HelmRelease called storageNamespace and let users pick their preferred storageNamespace

is a great idea, but in a way that is backwards compatible.

**Comment by gtracer** [comment]
Put up a draft PR to add the property. It is optional so I think it should maintain backwards compatibility while extending the functionality. 

Would love some input and if everyone is happy with the approach I can add some tests 

#198

**Comment by tommasopozzetti** [comment]
@gtracer I think this would be a perfect solution, solving all issues reported up to now here and maintaining backwards compatibility for current users.
I would suggest also changing the docs around helm-operator to helm-controller migration to point out that if users are using helm-operator with `helmreleases.helm.fluxcd.io/v1` exploiting the `targetNamespace` feature, they should convert those to `helmreleases.helm.toolkit.fluxcd.io/v2beta1` setting both `targetNamespace` and `storageNamespace` to the same namespace in order to maintain the same behavior.

**Comment by hiddeco** [comment]
Available in `helm-controller` `v0.6.1` and `flux` `v0.7.0`.
